core/src/main/java/io/substrait/expression/FieldReference.java

```diff
@@ -214,7 +214,10 @@ private static FieldReference of(
       if (i == 0) {
         var last = segments.get(0);
         reference =
-            type == null ? last.constructOnExpression(expression) : last.constructOnRoot(type);
+            type == null
+                ? last.constructOnExpression(expression)
+                : last.constructOnRoot(
+                    ((Type.Struct) type).fields().get(((StructField) last).offset()));
```

Comment:
We don't know the root type in advance, so we pass in the type of the input. The type of the field reference can then be computed based on indexing into the fields. Unclear what to do if `type` is not a struct.